src/riemann-client.1
.TH "RIEMANN\-CLIENT" "1" "March 2016" "The MadHouse Project" "riemann-c-client"

.SH "NAME"
\fBriemann\-client\fR \- Riemann communication tool

.SH "SYNOPSIS"
\fBriemann\-client\fR \fBquery\fR [\fIOPTIONS\fR...] \fIQUERY\fR [\fIHOST\fR] [\fIPORT\fR]
.br
\fBriemann\-client\fR \fBsend\fR [\fIOPTIONS\fR...] [\fIHOST\fR] [\fIPORT\fR]
.br
\fBriemann\-client\fR \fB\-\-help\fR|\fB\-?\fR
.br
\fBriemann\-client\fR \fB\-\-version\fR

.SH "DESCRIPTION"
\fBriemann\-client\fR is a very simple tool to facilitate communication
with a Riemann event stream processing server. It can be used both
for sending events to one, and to query it too.

.SH "OPTIONS"
In both \fBquery\fR and \fBsend\fR mode, \fIHOST\fR and \fIPORT\fR
default to \fIlocalhost\fR and \fI5555\fR, respectively.

.SS "Query mode"
In query mode, \fBriemann\-client\fR takes one mandatory argument: the
query itself. It must be a single argument, which will be sent to
Riemann as\-is, and the response dumped to the standard output, or, in
case of an error, the error message will be written to the standard
error.

.P
See the Riemann documentation \fIhttp://riemann.io/concepts.html\fR
for more information about the query syntax.

.P
Apart from the mandatory argument, the \fBriemann\-client\fR accepts
the following options in query mode:

.TP
\fB\-j\fR, \fB\-\-json\fR
Print the result in JSON format, instead of the default,
human\-readable one.

.TP
\fB\-T\fR, \fB\-\-tcp\fR
Send the event via TCP (the default).

.TP
\fB\-G\fR, \fB\-\-tls\fR
Send the event via TLS.

.TP
\fB\-o\fR, \fB\-\-option\fR \fIoption\fR=\fIvalue\fR
Set one client option to the given value.

Available client options are:

.RS
.TP
\fBcafile\fR
Path to the file containing trusted certificate authority
certificates. Only used for TLS connections.

.TP
\fBcertfile\fR
Path to the file containing the public client certificate. Only used
for TLS connections.

.TP
\fBkeyfile\fR
Path to the file containing the private client key. Only used for TLS
connections.

.TP
\fBpriorities\fR
Priorities for the cipher suites to be used for the session.
.RE

.SS "Submit mode"
In submit mode, \fBriemann\-client\fR accepts the following options:

.TP
\fB\-s\fR, \fB\-\-state\fR \fISTATE\fR
The state the event should describe, free\-form text, optional.

.TP
\fB\-S\fR, \fB\-\-service\fR \fISERVICE\fR
The service the event is coming from.

.TP
\fB\-h\fR, \fB\-\-host\fR \fIHOST\fR
The host the event originates from. Not to be confused with the
optional \fIHOST\fR parameter of \fBriemann\-client\fR itself, which
sets where to send the events to.

.TP
\fB\-D\fR, \fB\-\-description\fR \fIDESCRIPTION\fR
The description of the event, optional.

.TP
\fB\-t\fR, \fB\-\-tag\fR \fITAG\fR
This option can be used multiple times to add tags to the event.

.TP
\fB\-a\fR, \fB\-\-attribute\fR \fINAME\fR=\fIVALUE\fR
This option can be used multiple times to add custom attributes to the
event.

.TP
\fB\-i\fR, \fB\-\-metric\-sint64\fR \fIMETRIC\fR
Sets the metric for the event, using integer precision.

.TP
\fB\-d\fR, \fB\-\-metric\-d\fR \fIMETRIC\fR
Sets the metric of the event, using double precision.

.TP
\fB\-f\fR, \fB\-\-metric\-f\fR \fIMETRIC\fR
Sets the metric of the event, using floating point precision.

.TP
\fB\-L\fR, \fB\-\-ttl\fR \fITTL\fR
An optional time\-to\-live of the event, in seconds (floating point number).

.TP
\fB\-T\fR, \fB\-\-tcp\fR
Send the event via TCP (the default).

.TP
\fB\-U\fR, \fB\-\-udp\fR
Send the event via UDP.

.TP
\fB\-G\fR, \fB\-\-tls\fR
Send the event via TLS.

.TP
\fB\-o\fR, \fB\-\-option\fR \fIoption\fR=\fIvalue\fR
Set one client option to the given value.

Available client options are:

.RS
.TP
\fBcafile\fR
Path to the file containing trusted certificate authority
certificates. Only used for TLS connections.

.TP
\fBcertfile\fR
Path to the file containing the public client certificate. Only used
for TLS connections.

.TP
\fBkeyfile\fR
Path to the file containing the private client key. Only used for TLS
connections.

.TP
\fBpriorities\fR
Priorities for the cipher suites to be used for the session.
.RE

.TP
\fB\-0\fR,  \fB\-\-stdin\fR
Continuously read and send metrics and states from standard input. Uses
all other fields specified on the command line. For each line on
standard input, a new event will be sent to Riemann. If the input is a
number, it will be used as a metric. If not, it will be used as the
state. If it is a number, followed by a space and some string, then
both metric and state will be set for the outgoing event.

.SH "EXAMPLES"

.SS "Sending an event"

.nf
$ riemann\-client send \-\-state ok \e
                      \-\-service "client\-test" \e
                      \-\-host "my\-host" \e
                      \-\-description "A message from the Manual" \e
                      \-\-tag tag\-1 \-\-tag tag\-2 \e
                      \-\-attribute x\-manual=best\-friend \e
                      \-\-metric\-sint64 1 \e
                      \-\-ttl 300.5 \e
                      \-\-tcp
.fi

.SS "Sending metrics / state from standard input"

.nf
$ echo "1.5 ok" | \ \e
  riemann\-client send \-\-service "client\-test: stdin" \e
                      \-\-host "my\-host" \e
                      \-\-description "A message from STDIN" \e
                      \-\-tag stdin \e
                      \-\-tcp \e
                      \-\-stdin
.fi

.SS "Querying Riemann"

.nf
$ riemann\-client query 'host = "my\-host"'
.fi

.SS "Using TLS"

.nf
$ riemann\-client query 'host = "my\-host"' \e
                       \-\-tls \e
                       \-\-option cafile=tests/data/cacert.pem \e
                       \-\-option certfile=tests/data/client.crt \e
                       \-\-option keyfile=tests/data/client.key \e
                       \-\-option priorities=NORMAL \e
                       localhost 5554
.fi

.SH "AUTHOR"
Gergely Nagy \fIalgernon@madhouse\-project.org\fR


NEWS.md
riemann-c-client 1.10.3
=======================
Released on 2018-08-19

Bugfixes
--------

* Dependencies on the generated protobuf C sources have been fixed in the
  Makefile.
* A few harmless compile-time warnings have been silenced.
* Fixed a few typos in riemann-client(1).

riemann-c-client 1.10.2
=======================
Released on 2017-08-19

Bugfixes
--------

* A number of formatting issues in the API docs have been fixed.

riemann-c-client 1.10.1
=======================
Released on 2017-06-12

Bugfixes
--------

* The library compiles on OSX/Darwin again, after fixing a symbol aliasing issue.

  Reported by @ilovezfs.

riemann-c-client 1.10.0
=======================
Released on 2017-06-09

Features
--------

* Added support for micro-second timestamps (available since Riemann 0.2.13),
  implemented by Mathieu Corbin, @mcorbin.

Miscellaneous changes
---------------------

* The parts of the test suite that require a running Riemann server are now
  disabled by default, and require setting the `RCC_NETWORK_TESTS` environment
  variable to a non-zero value, before running the test suite.

  Fixes an issue on systems that had something else listening on the port, or a
  Riemann with a different config than the one the test suite expected.

  Reported by Dave Cottlehuber, @dch.

riemann-c-client 1.9.1
======================
Released on 2016-09-30

Bugfixes
--------

* Fix the build on non-ELF systems, and systems that lack symbol versioning.
  (Thanks to @ilovezfs)
* Use the `PROTOBUF_C_CFLAGS` environment variable, when set.
  (Thanks to @ciomaire)
* Added some missing includes that prevented the library from compiling on
  FreeBSD.
  (Thanks to Dave Cottlehuber, @dch)
* Fixed an issue with how some of the symbols were versioned, which caused link
  failures in some situations.
  (Thanks to Dave Cottlehuber, @dch)

riemann-c-client 1.9.0
======================
Released on 2016-04-25

Features
--------

* Added the `RIEMANN_CLIENT_OPTION_TLS_PRIORITIES` client option, which can be
  used to set priorities for cipher suites to be used for a TLS session.

  The `riemann-client` tool also learned to accept a `-o priorities=` option,
  and set the above one appropriately.

riemann-c-client 1.8.1
======================
Released on 2015-08-28

Bugfixes
--------

* Tests use a different port for connection refused tests

  Because port 5557 is the default for (repl-server), use 5559 for
  testing for library behaviour in the face of connection
  refusal. This is to make local testing without a special config
  easier.

* Fixed a compile-time warning when compiling with GnuTLS 3.2

  When compiling with TLS support enabled, using GnuTLS 3.2, there was
  a harmless compile-time warning, which has now been corrected.

Miscellaneous changes
---------------------

* Completely new documentation

  Instead of using Doxygen to document our API, a new, prosaic
  documentation was written, that explains not only the hows, but the
  whys too. It is hoped that the new documentation is more
  approachable, and explains the library better.

riemann-c-client 1.8.0
======================
Released on 2015-06-04

Features
--------

* Added the `riemann_client_set_timeout` function.

  To be able to reliably detect whether a connection died, we need to
  set a timeout for the blocking operations. The new
  `riemann_client_set_timeout()` function does just that.

  Suggested by Fabien Wernli and others.

* Added a set of version macros.

  The `RCC_MAJOR_VERSION`, `RCC_MINOR_VERSION`, `RCC_PATCH_VERSION`
  and `RCC_VERSION_NUMBER` macros were added to aid developers in
  determining the version of the library at build-time without the
  need of `pkg-config`.

  The primary reason to do this would be to use features of the
  library, if available, optionally.

* The `riemann-client` utility can now forward messages from STDIN

  The `riemann-client` utility gained an `--stdin` option. Using this
  option, one can set up a template with the command line options, and
  send multiple events with different metrics and/or states, by piping
  lines to the utility's standard input.

  See the manual page for more information and an example.

Bugfixes
--------

* Handshake timeouts now work with GnuTLS 2.x too.

  When compiled with GnuTLS 2.x, the
  RIEMANN_CLIENT_OPTION_TLS_HANDSHAKE_TIMEOUT option was silently
  ignored. Now the timeouts are properly set up, but the timeout
  applies to all TLS operations, not only the handshake.

riemann-c-client 1.7.0
======================
Released on 2015-05-04

Features
--------

* Added the `riemann_communicate_query` and
  `riemann_communicate_event` functions.

  To make the `riemann_communicate()` API simpler for common use
  cases, two new functions were introduced, to be able to easily send
  queries and receive replies, and another to send a single event, and
  read the ACK back.

Bugfixes
--------

* Fix the TLS support to work with GnuTLS 3.1 tool

  Treat anything less than GnuTLS 3.3 as if it were GnuTLS 2.x, so
  that the library compiles and works for every possible GnuTLS
  version between 2.8 and 3.3+.

  Reported by Peter Czanik.

* Add a TLS example to the riemann-client(1) manual page.

  Suggested by Fabien Wernli.

riemann-c-client 1.6.1
======================
Released on 2015-04-23

Bugfixes
--------

* TLS support's GnuTLS dependency lowered even more

  Turns out that we need a little more work to support GnuTLS 2.8, and
  the previous release only lowered the requirement to 2.10+. With
  this release, 2.8+ is supported too.

  Reported by Fabien Wernli.

* Allow compiling with GCC < 4.4

  A #pragma used in the TLS code did not work with gcc < 4.4, this has
  been corrected.

  Reported by Fabien Wernli.

riemann-c-client 1.6.0
======================
Released on 2015-04-23

Features
--------

* TLS support's GnuTLS dependency lowered

  The minimum version of GnuTLS required was lowered from 3.3 to 2.8,

src/riemann-client.c
/* riemann/riemann-client.c -- Riemann C client library
 * Copyright (C) 2013-2017  Gergely Nagy <algernon@madhouse-project.org>
 *
 * This library is free software: you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public License
 * as published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

#include <riemann/riemann-client.h>

#include <errno.h>
#include <inttypes.h>
#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <getopt.h>
#include <time.h>

#include "riemann/platform.h"

#if HAVE_JSON_C
#include <json.h>
#endif

static int
version_display (int __attribute__((unused)) argc, char __attribute__((unused)) *argv[])
{
  printf ("%s\n", riemann_client_version_string ());
  return EXIT_SUCCESS;
}

static void
help_display (const char *app_name, void (*contents)(void))
{
  version_display (0, NULL);
  printf ("Usage: %s COMMAND [options...] [HOST] [PORT]\n"
          "\n", app_name);
  contents();
  printf ("\n"
          "The HOST and PORT arguments are optional, and they default to\n"
          "\"localhost\" and 5555, respectively.\n"
          "\n"
          "Report " PACKAGE_NAME " bugs to " PACKAGE_BUGREPORT "\n");
}

#include "cmd-send.c"
#include "cmd-query.c"

static void
help_generic (void)
{
  printf ("Available commands: send and query.\n"
          "\n");
  help_send ();
  printf ("\n");
  help_query ();
}

static int
display_help_generic (int __attribute__((unused)) argc,
                      char *argv[])
{
  help_display (argv[0], help_generic);
  return EXIT_SUCCESS;
}

typedef int (*cmd_t) (int argc, char *argv[]);

static struct
{
  const char *command;
  cmd_t handler;
  int (*cmp) (const char *s1, const char *s2);
} command_map[] = {
  { "send", client_send, strcasecmp },
  { "query", client_query, strcasecmp },
  { "--version", version_display, strcasecmp },
  { "-V", version_display, strcmp },
  { "--help", display_help_generic, strcasecmp },
  { "-?", display_help_generic, strcmp },
  { NULL, NULL, NULL },
};

int
main (int argc, char *argv[])
{
  const char *command = NULL;
  int i = 0;

  if (argc < 2)
    {
      fprintf (stderr, "Not enough arguments!\n");
      display_help_generic (argc, argv);
      exit (EXIT_FAILURE);
    }

  command = argv[1];

  while (command_map[i].command != NULL)
    {
      if (command_map[i].cmp (command, command_map[i].command) == 0)
        return command_map[i].handler (argc, argv);
      i++;
    }

  fprintf (stderr, "Unknown command: '%s'\n", command);
  help_display (argv[0], help_generic);
  return EXIT_FAILURE;
}
